### DISCUSSION TITLE
Caches is not defined  at WebviewResourceCache.resolveCache

### DISCUSSION THREAD

**Discussion by wss29** [eclipse-theia/theia#14197] [Q&A]
Hi,
   I installed a plugin and ran it theia web browser, which is [vscode-pdf](https://open-vsx.org/extension/tomoki1207/pdf) used to view pdf in the web browser in theia.
  When I run my app in the browser and access it with localhost:3000, the plugin works ok, I can open pdf in the browser. but when I run my app in the browser with my computer IP like 192.168.1.120:3000, I get an error about the plugin, it seems that some error happened in plugin-ext
![image](https://github.com/user-attachments/assets/39e6d50a-9fed-402f-b2b4-f3319e8004d1)
![image](https://github.com/user-attachments/assets/4601117b-4cf7-40f4-9a18-ca9ada1bb1ba)
  Does anyone have experience with this problem?

**Comment by msujew** [comment]
Hey @wss29,

the [`caches`](https://developer.mozilla.org/en-US/docs/Web/API/Window/caches) global variable is only available in secure contexts (i.e. HTTPS). This is expected behavior when accessing an IP without a HTTPS/TLS certificate.

**Reply by wss29** [reply]
Hi @msujew
 Does that means I can not use [vscode-pdf](https://open-vsx.org/extension/tomoki1207/pdf) in theia browser app? Can I do something to make it work in browser? vscode-pdf seems the best choice for pdf view currently for theia

**Reply by msujew** [reply]
You can, but you need a secure context. I.e. a security certificate for your website. Aside from that, I don't think that the error actually prevents you from using the extension, right? It only logs that it cannot perform the caching, which isn't relevant for usage of the pdf viewer.

**Reply by wss29** [reply]
Hi @msujew 
thank you for your reply. In fact, when this error happens, I can not open pdf file in my browser app. you says security certificate means add security certificate to my theia web app and I can access my app like https:192.168.1.120:3000?

**Reply by msujew** [reply]
Yes, but I don't think it's actually possible to get a TLS certificate for that IP range. For any other IP, see https://www.cloudflare.com/learning/ssl/what-is-an-ssl-certificate/.

**Reply by wss29** [reply]
Hi @msujew 
  My web app only runs on a local area network， not connected to the internet. I search for some solutions, for example, using OpenSSL
![image](https://github.com/user-attachments/assets/93811358-b9b1-4655-993d-68da42d5f375)
  but it needs to modify the server code as follows
![image](https://github.com/user-attachments/assets/9d578aa5-e8f4-45e4-b8eb-519ae876fc77)
 Does theia support support this kind of operation?

**Reply by msujew** [reply]
You can start Theia with the `ssl`, `cert` and `certkey` arguments to set those (you can also run `theia start --help` to get more info on those arguments).

**Reply by wss29** [reply]
Hi @msujew 
After setting ssl, cert and certkey, the cache error does not appear, but I still can not open pdf in browser, have you ever used this kind of plugin like [vscode-pdf](https://open-vsx.org/extension/tomoki1207/pdf) or [vscode-office](https://open-vsx.org/extension/cweijan/vscode-office) to view docx or pdf in browser of theia? It is very strange, why  this does not work
